tests/forms/test_address_upload.py

```diff
@@ -0,0 +1,170 @@
+    assert WalletAddress.objects.filter(hex_address=existing_wallet_address).exists()
+
+    client.login(email='admin@pretix', password='admin')
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    response = client.get(request_url)
+
+    assert response.status_code == 200
+    assert b'Wallet address upload' in response.content
+
+    file = SimpleUploadedFile('upload.txt', f"""
+        # OK
+        0x38A670EB58F937D63D79D92dC651155595e66009
+        # OK
+        0x26AA4d587584FE1f32707070768236eDC625dDD7
+        # Duplicate
+        0x26AA4d587584FE1f32707070768236eDC625dDD7
+        # Commented
+        # 0x0a648dD253CfBf0A7340C93fe7D8774FEfe972eD
+        # Already exists in database
+        {existing_wallet_address}
+        # Weird spacing
+            0x06E5c4854D7A21A4C466bcA74F6A2867a8ef0fd2
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': file}, follow=True)
+
+    assert response.status_code == 200
+    assert b'5 total addresses' in response.content
+    assert b'4 unique addresses' in response.content
+    assert b'1 addresses that already exist in the database' in response.content
+    assert b'3 new addresses that don\'t already exist in the database' in response.content
+
+    response = client.post(request_url + 'confirm/', {'action': 'confirm'}, follow=True)
+
+    assert response.status_code == 200
+    assert b'Created 3 new wallet addresses!' in response.content
+
+    loaded_addresses = [
+        '0x38A670EB58F937D63D79D92dC651155595e66009',
+        '0x26AA4d587584FE1f32707070768236eDC625dDD7',
+        '0x06E5c4854D7A21A4C466bcA74F6A2867a8ef0fd2'
+    ]
+
+    for address in loaded_addresses:
+        assert WalletAddress.objects.filter(hex_address=address).exists()
+
+
+@pytest.mark.django_db
+def test_file_upload_length_short(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        0x38A670EB58F9D63D79D92dC651155595e66009
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
+    assert b'Syntax error on line 1:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_length_long(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        0x38A670EB58F9D63Dfds32D279D92dC651155595e66009
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
```

Comment:
I'm not *sure* this is actually wrong as it's been a while since I did a lot of "web" and my standard status code knowledge is rusty but the 2xx codes tend to mean "success".

Some quick googling suggests we should be using something in the 4xx range: https://stackoverflow.com/questions/47269601/what-http-response-code-to-use-for-failed-post-request

I don't think this is something you need to cleanup in this PR (avoid scope creep), but it might be a nice thing to open an issue for?

Reply:
Opened new issue #125